File: NotePal/Sources/NotePal/Stores/WellnessReminderStore.swift
import Foundation

@MainActor
final class WellnessReminderStore: ObservableObject {
    @Published private(set) var reminders: [WellnessReminder] = []
    @Published var lastError: String?

    var onRemindersChanged: (() -> Void)?
    var onError: ((String) -> Void)?

    private let storage: LocalDataStore

    init(storage: LocalDataStore = .shared) {
        self.storage = storage
        reload()
    }

    func reload() {
        do {
            let data = try storage.update { data in
                normalizeWellnessData(in: &data)
            }
            reminders = sort(data.wellnessReminders)
            lastError = nil
            onRemindersChanged?()
        } catch {
            lastError = "无法加载养生提醒：\(error.localizedDescription)"
            onError?(lastError ?? "无法加载养生提醒。")
        }
    }

    func setEnabled(kind: WellnessReminder.Kind, isEnabled: Bool) {
        persist { data in
            normalizeWellnessData(in: &data)
            guard let index = data.wellnessReminders.firstIndex(where: { $0.kind == kind }) else {
                return
            }

            data.wellnessReminders[index].isEnabled = isEnabled
            data.wellnessReminders[index].updatedAt = Date()
            if isEnabled {
                let interval = TimeInterval(data.wellnessReminders[index].effectiveIntervalMinutes * 60)
                data.wellnessReminders[index].nextReminderAt = Date().addingTimeInterval(interval)
            }
        }
    }

    func setInterval(kind: WellnessReminder.Kind, minutes: Int) {
        persist { data in
            normalizeWellnessData(in: &data)
            guard let index = data.wellnessReminders.firstIndex(where: { $0.kind == kind }) else {
                return
            }

            let normalizedMinutes = min(max(minutes, 5), 1440)
            data.wellnessReminders[index].intervalMinutes = normalizedMinutes
            data.wellnessReminders[index].nextReminderAt = Date().addingTimeInterval(TimeInterval(normalizedMinutes * 60))
            data.wellnessReminders[index].updatedAt = Date()
        }
    }

    func dueReminders(asOf date: Date = Date()) -> [WellnessReminder] {
        reminders.filter { reminder in
            reminder.isEnabled && reminder.nextReminderAt <= date
        }
    }

    func nextDueDate(asOf date: Date = Date()) -> Date? {
        reminders
            .filter { $0.isEnabled && $0.nextReminderAt > date }
            .map(\.nextReminderAt)
            .min()
    }

    func advanceReminders(_ dueReminders: [WellnessReminder], from date: Date = Date()) {
        guard !dueReminders.isEmpty else {
            return
        }

        let dueKinds = Set(dueReminders.map(\.kind))
        persist { data in
            normalizeWellnessData(in: &data)
            for index in data.wellnessReminders.indices where dueKinds.contains(data.wellnessReminders[index].kind) {
                let interval = TimeInterval(data.wellnessReminders[index].effectiveIntervalMinutes * 60)
                data.wellnessReminders[index].nextReminderAt = date.addingTimeInterval(interval)
                data.wellnessReminders[index].updatedAt = date
            }
        }
    }

    private func persist(_ mutation: (inout NotePalData) -> Void) {
        do {
            let data = try storage.update(mutation)
            reminders = sort(data.wellnessReminders)
            lastError = nil
            onRemindersChanged?()
        } catch {
            lastError = "无法保存养生提醒：\(error.localizedDescription)"
            onError?(lastError ?? "无法保存养生提醒。")
        }
    }

    private func sort(_ reminders: [WellnessReminder]) -> [WellnessReminder] {
        let order = Dictionary(uniqueKeysWithValues: WellnessReminder.Kind.allCases.enumerated().map { ($1, $0) })
        return reminders.sorted { first, second in
            (order[first.kind] ?? 0) < (order[second.kind] ?? 0)
        }
    }
}

private func normalizeWellnessData(in data: inout NotePalData) {
    let now = Date()
    var remindersByKind = Dictionary(uniqueKeysWithValues: WellnessReminder.defaults(now: now).map { ($0.kind, $0) })

    for reminder in data.wellnessReminders {
        var normalized = reminder
        normalized.intervalMinutes = normalized.effectiveIntervalMinutes
        remindersByKind[normalized.kind] = normalized
    }

    // One-time migration from the earlier MVP, where wellness reminders were
    // represented as ordinary todos. They are now independent timed reminders.
    for legacyTodo in data.todos where legacyTodo.isWellnessTodo {
        guard
            let rawKind = legacyTodo.wellnessKind,
            let kind = WellnessReminder.Kind(rawValue: rawKind)
        else {
            continue
        }

        let intervalMinutes = legacyTodo.effectiveWellnessIntervalMinutes
        remindersByKind[kind] = WellnessReminder(
            kind: kind,
            isEnabled: !legacyTodo.isCompleted,
            intervalMinutes: intervalMinutes,
            nextReminderAt: legacyTodo.dueDate ?? now.addingTimeInterval(TimeInterval(intervalMinutes * 60)),
            updatedAt: legacyTodo.updatedAt
        )
    }

    data.todos.removeAll { $0.isWellnessTodo }
    data.wellnessReminders = WellnessReminder.Kind.allCases.compactMap { remindersByKind[$0] }

    let currentTodoReminderKeys = Set(data.todos.map(\.reminderRevisionKey))
    data.remindedTodoRevisions = data.remindedTodoRevisions.intersection(currentTodoReminderKeys)
}

File: NotePal/Windows/NotePal.Windows/Services/WellnessReminderStore.cs
using NotePalWindows.Models;

namespace NotePalWindows.Services;

public sealed class WellnessReminderStore
{
    private readonly LocalDataStore _dataStore;

    public WellnessReminderStore(LocalDataStore dataStore)
    {
        _dataStore = dataStore;
        Reload();
    }

    public event EventHandler? Changed;
    public Action<string>? ErrorRaised { get; set; }
    public string? LastError { get; private set; }
    public IReadOnlyList<WellnessReminder> Reminders { get; private set; } = [];

    public void Reload()
    {
        try
        {
            var data = _dataStore.Update(LocalDataStore.Normalize);
            Reminders = Sort(data.WellnessReminders);
            ClearError();
        }
        catch (Exception ex)
        {
            SetError($"无法加载养生提醒：{ex.Message}");
        }
    }

    public void SetEnabled(WellnessReminderKind kind, bool isEnabled)
    {
        Persist(data =>
        {
            var reminder = EnsureReminder(data, kind);
            reminder.IsEnabled = isEnabled;
            reminder.UpdatedAt = DateTimeOffset.Now;
            if (isEnabled)
            {
                reminder.NextReminderAt = DateTimeOffset.Now.AddMinutes(reminder.EffectiveIntervalMinutes);
            }
        });
    }

    public void SetInterval(WellnessReminderKind kind, int minutes)
    {
        Persist(data =>
        {
            var reminder = EnsureReminder(data, kind);
            reminder.IntervalMinutes = Math.Clamp(minutes, 5, 1440);
            reminder.NextReminderAt = DateTimeOffset.Now.AddMinutes(reminder.EffectiveIntervalMinutes);
            reminder.UpdatedAt = DateTimeOffset.Now;
        });
    }

    public IReadOnlyList<WellnessReminder> DueReminders(DateTimeOffset asOf)
    {
        return Reminders
            .Where(reminder => reminder.IsEnabled && reminder.NextReminderAt <= asOf)
            .ToList();
    }

    public DateTimeOffset? NextDueDate(DateTimeOffset asOf)
    {
        return Reminders
            .Where(reminder => reminder.IsEnabled && reminder.NextReminderAt > asOf)
            .Select(reminder => (DateTimeOffset?)reminder.NextReminderAt)
            .Min();
    }

    public void AdvanceReminders(IEnumerable<WellnessReminder> reminders, DateTimeOffset from)
    {
        var kinds = reminders.Select(reminder => reminder.Kind).ToHashSet();
        if (kinds.Count == 0)
        {
            return;
        }

        Persist(data =>
        {
            foreach (var reminder in data.WellnessReminders.Where(reminder => kinds.Contains(reminder.Kind)))
            {
                reminder.NextReminderAt = from.AddMinutes(reminder.EffectiveIntervalMinutes);
                reminder.UpdatedAt = from;
            }
        });
    }

    private void Persist(Action<NotePalData> mutation)
    {
        try
        {
            var data = _dataStore.Update(data =>
            {
                LocalDataStore.Normalize(data);
                mutation(data);
            });
            Reminders = Sort(data.WellnessReminders);
            ClearError();
        }
        catch (Exception ex)
        {
            SetError($"无法保存养生提醒：{ex.Message}");
        }
    }

    private static WellnessReminder EnsureReminder(NotePalData data, WellnessReminderKind kind)
    {
        LocalDataStore.Normalize(data);
        var reminder = data.WellnessReminders.FirstOrDefault(item => item.Kind == kind);
        if (reminder is not null)
        {
            return reminder;
        }

        reminder = new WellnessReminder { Kind = kind };
        data.WellnessReminders.Add(reminder);
        return reminder;
    }

    private static IReadOnlyList<WellnessReminder> Sort(IEnumerable<WellnessReminder> reminders)
    {
        var order = Enum.GetValues<WellnessReminderKind>()
            .Select((kind, index) => new { kind, index })
            .ToDictionary(item => item.kind, item => item.index);

        return reminders
            .OrderBy(reminder => order.TryGetValue(reminder.Kind, out var index) ? index : 0)
            .Select(reminder => reminder.Clone())
            .ToList();
    }

    private void ClearError()
    {
        LastError = null;
        Changed?.Invoke(this, EventArgs.Empty);
    }

    private void SetError(string message)
    {
        LastError = message;
        ErrorRaised?.Invoke(message);
        Changed?.Invoke(this, EventArgs.Empty);
    }
}
